mono/profiler/proflog.c

```diff
@@ -1104,85 +1225,145 @@ static uint64_t last_hs_time = 0;
 static void
 heap_walk (MonoProfiler *profiler)
 {
-	int do_walk = 0;
-	uint64_t now;
-	LogBuffer *logbuffer;
 	if (!do_heap_shot)
 		return;
-	logbuffer = ensure_logbuf (
-		EVENT_SIZE /* event */ +
-		LEB128_SIZE /* time */
-	);
-	now = current_time ();
-	if (hs_mode_ms && (now - last_hs_time)/1000000 >= hs_mode_ms)
-		do_walk = 1;
+
+	gboolean do_walk = 0;
+	uint64_t now = current_time ();
+
+	if (hs_mode_ms && (now - last_hs_time) / 1000000 >= hs_mode_ms)
+		do_walk = TRUE;
 	else if (hs_mode_gc && (gc_count % hs_mode_gc) == 0)
-		do_walk = 1;
+		do_walk = TRUE;
 	else if (hs_mode_ondemand)
 		do_walk = heapshot_requested;
 	else if (!hs_mode_ms && !hs_mode_gc && profiler->last_gc_gen_started == mono_gc_max_generation ())
-		do_walk = 1;
+		do_walk = TRUE;
 
 	if (!do_walk)
 		return;
+
 	heapshot_requested = 0;
+
+	ENTER_LOG;
+
+	LogBuffer *logbuffer = ensure_logbuf (
+		EVENT_SIZE /* event */ +
+		LEB128_SIZE /* time */
+	);
+
 	emit_byte (logbuffer, TYPE_HEAP_START | TYPE_HEAP);
 	emit_time (logbuffer, now);
+
 	mono_gc_walk_heap (0, gc_reference, NULL);
+
 	logbuffer = ensure_logbuf (
 		EVENT_SIZE /* event */ +
 		LEB128_SIZE /* time */
 	);
+
 	now = current_time ();
+
 	emit_byte (logbuffer, TYPE_HEAP_END | TYPE_HEAP);
 	emit_time (logbuffer, now);
+
+	EXIT_LOG;
+
 	last_hs_time = now;
 }
 
 static void
-gc_event (MonoProfiler *profiler, MonoGCEvent ev, int generation) {
-	uint64_t now;
+gc_event (MonoProfiler *profiler, MonoGCEvent ev, int generation)
+{
+	ENTER_LOG;
+
 	LogBuffer *logbuffer = ensure_logbuf (
 		EVENT_SIZE /* event */ +
 		LEB128_SIZE /* time */ +
 		LEB128_SIZE /* gc event */ +
 		LEB128_SIZE /* generation */
 	);
-	now = current_time ();
-	ENTER_LOG (logbuffer, "gcevent");
+
+	uint64_t now = current_time ();
+
 	emit_byte (logbuffer, TYPE_GC_EVENT | TYPE_GC);
 	emit_time (logbuffer, now);
 	emit_value (logbuffer, ev);
 	emit_value (logbuffer, generation);
-	/* to deal with nested gen1 after gen0 started */
-	if (ev == MONO_GC_EVENT_START) {
+
+	EXIT_LOG;
+
+	switch (ev) {
+	case MONO_GC_EVENT_START:
+		/* to deal with nested gen1 after gen0 started */
 		profiler->last_gc_gen_started = generation;
+
 		if (generation == mono_gc_max_generation ())
 			gc_count++;
-	}
-	if (ev == MONO_GC_EVENT_PRE_START_WORLD)
+		break;
+	case MONO_GC_EVENT_PRE_STOP_WORLD:
+		/*
+		 * Ensure that no thread can be in the middle of writing to
+		 * a buffer when the world stops...
+		 */
+		buffer_lock_excl ();
```

Comment:
threads can be suspended outside of STW, if a thread is suspended while holding the shared lock we can run into trouble. In theory, at least.